limiter/limiter.go

```diff
@@ -0,0 +1,274 @@
+// a group from some on-disk database, so you don't have to have all group
+// limits in memory. (Limiter itself will clear out unused groups from its own
+// memory.)
+type SetLimitCallback func(name string) int
+
+// Limiter struct is used to limit usage of groups.
+type Limiter struct {
+	cb     SetLimitCallback
+	groups map[string]*group
+	mu     sync.Mutex
+}
+
+// New creates a new Limiter.
+func New(cb SetLimitCallback) *Limiter {
+	return &Limiter{
+		cb:     cb,
+		groups: make(map[string]*group),
+	}
+}
+
+// SetLimit creates or updates a group with the given limit.
+func (l *Limiter) SetLimit(name string, limit uint) {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	if g, set := l.groups[name]; set {
+		g.setLimit(limit)
+	} else {
+		l.groups[name] = newGroup(name, limit)
+	}
+}
+
+// GetLimit tells you the limit currently set for the given group. If the group
+// doesn't exist, returns -1.
+func (l *Limiter) GetLimit(name string) int {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	group := l.vivifyGroup(name)
+	if group == nil {
+		return -1
+	}
+
+	return int(group.limit)
+}
+
+// RemoveLimit removes the given group from memory. If your callback also begins
+// returning -1 for this group, the group effectively becomes unlimited.
+func (l *Limiter) RemoveLimit(name string) {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	delete(l.groups, name)
+}
+
+// Increment sees if it would be possible to increment the count of every
+// supplied group, without making any of them go over their limit.
+//
+// If this is the first time we're seeing a group name, or a Decrement() call
+// has made us forget about that group, the callback provided to New() will be
+// called with the name, and the returned value will be used to create a new
+// group with that limit and initial count of 0 (which will become 1 if this
+// returns true). Groups with a limit of 0 will not be able to be Increment()ed.
+//
+// If possible, the group counts are actually incremented and this returns
+// true. If not possible, no group counts are altered and this returns false.
+//
+// If an optional wait duration is supplied, will wait for up to the given wait
+// period for an increment of every group to be possible.
+func (l *Limiter) Increment(groups []string, wait ...time.Duration) bool {
+	l.mu.Lock()
+	if l.checkGroups(groups) {
+		l.incrementGroups(groups)
+		l.mu.Unlock()
+
+		return true
+	}
+
+	if len(wait) != 1 {
+		l.mu.Unlock()
+
+		return false
+	}
+
+	ch := make(chan bool, len(groups))
+	l.registerGroupNotifications(groups, ch)
+	l.mu.Unlock()
+
+	limit := time.After(wait[0])
+
+	return waitForIncrementOfEveryGroup(limit, l, groups, ch)
+}
+
+// waitForIncrementOfEveryGroup is a sub function of Increment which waits for up to the given wait
+// period for an increment of every group to be possible.
+func waitForIncrementOfEveryGroup(limit <-chan time.Time, l *Limiter, groups []string, ch chan bool) bool {
+	for {
+		select {
+		case <-ch:
+			l.mu.Lock()
```

Comment:
This check, increment and then register block is the same as above. Extract in to a method regardless, and then possibly only call it from one place if you also refactor an efficient way to always wait on the context.